api/README.md

```diff
@@ -1 +1,269 @@
+
+#### Should the endpoint name be singular or plural?
+The keep-it-simple rule applies here. Although your inner-grammatician will tell you it's wrong to describe a single instance of a resource using a plural, the pragmatic answer is to keep the URL format consistent and always use a **plural**.
+
+### Relations
+
+If a resource can only exist within another resource these are the RESTful principles you can use:
+
+- `GET /posts/31/comments`        - Retrieves a list of comments for post with id 31.
+- `POST /posts/31/comments`       - Creates a new comment on post with id 31.
+- `GET /posts/31/comments/25`     - Retrieves the comment with id 25 on post with id 31.
+- `PUT /posts/31/comments/25`     - Updates the comment with id 25 on post with id 31.
+- `PATCH /posts/31/comments/25`   - Partially updates the comment with id 25 on post with id 31.
+- `DELETE /posts/31/comments/25`  - Deletes the comment with id 25 on post with id 31.
+
+If a resource can exist independently it's advised to just include the resource's identifier in the parent. Alternatively if the client needs multiple related resources at the same time they should be bundled together into the same response to reduce the number of requests.
+
+### Actions
+
+After the resources are defined you need to find out what kind of actions should be applied to them. The basic CRUD actions can be omitted by simply using the HTTP request methods mentioned above. If an action doesn't fit into the CRUD operations there are two main approaches you can choose from:
+
+1. Treat it like a sub-resource and modify it with CRUD operations. E.g. `PUT /posts/31/like` to like the post and `DELETE /posts/31/like` to dislike the post.
+1. There are cases when you can't map the action to a single resource. In this case naming the endpoint with the name of the action makes the most sense, e.g. `/search`, `/sign-in`, etc.
+
+**You should never ever** leak any implementation details via your API.
+
+## HTTP Request Methods
+
+The HTTP request methods `GET`, `POST`, `PUT`, `PATCH` and `DELETE` have specific meaning in manipulating these.
+
+### GET
+
+- Retrieves a resource or a list of resources.
+- It should **never** update a resource.
+- Shouldn't even update the access logs.
+- Use query strings instead of request body.
+
+To retrieve the list of posts filtered by the keyword "awesome":
+```HTTP
+GET /posts?search=awesome
+```
+
+To retrieve the post with id 31:
+```HTTP
+GET /posts/31
+```
+
+### POST
+
+- Creates a resource.
+- It should be used when the server has to generate a unique key for the created resource.
+- Multiple POST requests to the same URL create duplicate records.
+
+To create a post:
+```HTTP
+POST /posts
+Request payload:
+{
+  "title": "Awesome title",
+  "content": "even more awesomeness"
+}
+```
+
+> The newly created resource (with a unique identifier) should be returned in the response body.
```

Comment:
Every resource should contain some basic auditing fields like create/update date which are generated upon creating/updating the entry. When the client doesn't require these fields the response can be omitted by returning 204 No Content. The best practice should be to wait for the API's response and process that. This way the API remains generic, decoupled from the client. By sending back a partial resource, the API would force all the clients to implement a specific logic to work with these resources (setting the id of the new item and inserting in the list instead of inserting a new item which is sent back by the server)